pandas/_libs/lib.pyx

```diff
@@ -2875,10 +2874,59 @@ def map_infer_mask(
     mask : ndarray
         uint8 dtype ndarray indicating values not to apply `f` to.
     convert : bool, default True
-        Whether to call `maybe_convert_objects` on the resulting ndarray
+        Whether to call `maybe_convert_objects` on the resulting ndarray.
     na_value : Any, optional
         The result value to use for masked values. By default, the
-        input value is used
+        input value is used.
+    dtype : numpy.dtype
+        The numpy dtype to use for the result ndarray.
+
+    Returns
+    -------
+    np.ndarray
+    """
+    # Passed so we can use infused types depending on the result dtype
+    dummy = np.empty(0, dtype=dtype)
+    result = _map_infer_mask(
+        arr,
+        f,
+        mask,
+        dummy,
+        convert,
+        na_value,
+        dtype,
+    )
+    if convert:
+        return maybe_convert_objects(result)
+    else:
+        return result
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _map_infer_mask(
+        ndarray[object] arr,
+        object f,
+        const uint8_t[:] mask,
+        numeric_object_t[:] dummy,
+        bint convert=True,
+        object na_value=no_default,
+        cnp.dtype dtype=np.dtype(object)
+) -> np.ndarray:
+    """
```

Comment:
NBD but i dont think we need to duplicate the docstring

Reply:
You're asking for the top line of the docstring for _map_infer_mask to be an exact duplicate of the one on map_infer_mask?

Reply:
Planning to just leave the first line - but can remove that too if preferred.